ObserverSet.swift

```diff
@@ -26,122 +26,198 @@
-   */
-  open func notify(_ parameters: Parameters) {
-    var callBackList: [(Parameters) -> Void] = []
-    synchronized {
-      for entry in self.entries {
-        if let observer = entry.observer {
-          callBackList.append(entry.callBack(observer))
-        }
-      }
-      self.entries = self.entries.filter{ $0.observer != nil }
-    }
-    for callBack in callBackList {
-      callBack(parameters)
-    }
-  }
-  
-  // MARK: - Private functions
-  // MARK: Locking support
-  
-  private var queue = DispatchQueue(label: "com.theappbusiness.ObserverSet", attributes: [])
-  
-  private func synchronized(_ f: () -> Void) {
-    queue.sync(execute: f)
-  }
+/// A set of observers that can be notified of certain actions. A more Swift-like version of NSNotificationCenter.
+open class ObserverSet<Parameters> {
+	
+	// MARK: - Private properties
+	
+	private var entries: [ObserverSetEntry<Parameters>] = []
+	
+	// MARK: - Initialisers
+	
+	/**
+	Creates a new instance of an observer set.
+	
+	- returns: A new instance of an observer set.
+	*/
+	public init() {}
+	
+	// MARK: - Public functions
+	
+	/**
+	Adds a new observer to the set.
+	
+	- parameter queue: The queue to call the function on when the notification is delivered. If nil (the default) it will be delivered on the same queue the notification was sent.
+	- parameter observer: The object that is to be notified.
+	- parameter callBack: The function to call on the observer when the notification is to be delivered.
+	
+	- returns: An entry in the list of observers, which can be used later to remove the observer.
+	*/
+	@discardableResult
+	open func add<ObserverType: AnyObject>(queue: DispatchQueue? = nil, _ observer: ObserverType, _ callBack: @escaping (ObserverType) -> (Parameters) -> Void) -> ObserverSetEntry<Parameters> {
+		let entry = ObserverSetEntryImpl<ObserverType, Parameters>(queue: queue, observer: observer, callBack: callBack)
+		synchronized {
+			self.entries.append(entry)
+		}
+		return entry
+	}
+	
+	/**
+	Adds a new function (or closure) to call when a notification is to be delivered.
+	This callback will be notified until it is removed by passing the return value to the `remove(_:)` method
+	
+	The return value is not discardable as the callback should be removed manually. To attach it to the life cycle of an object, use the `add` method that takes a `lifeCycleObserver` parameter.
+	
+	- parameter queue: The queue to call the function on when the notification is delivered. If nil (the default) it will be delivered on the same queue the notification was sent.
+	- parameter callBack: The function to call when the notification is to be delivered.
+	
+	- returns: An entry in the list of observers, which should be used later to remove the observer.
+	*/
+	open func add(queue: DispatchQueue? = nil, _ callBack: @escaping (Parameters) -> Void) -> ObserverSetEntry<Parameters> {
```

Comment:
I'm not 100% sure about this but it makes sense that this shouldn't be @discardableResult to prompt the user to remember to remove the observer at some point. If they really want it to be always added then they can do `_ = observerSet.add(...)`  which I've had to do in some places in the Tesco App where we set up observers on app launch.